Objects/unicodeobject.c

```diff
@@ -14617,6 +14617,62 @@ unicode_new_impl(PyTypeObject *type, PyObject *x, const char *encoding,
     return unicode;
 }
 
+static const char *
+arg_as_utf8(PyObject *obj, const char *name)
+{
+    if (!PyUnicode_Check(obj)) {
+        PyErr_Format(PyExc_TypeError,
+                     "str() argument '%s' must be str, not %T",
+                     name, obj);
+        return NULL;
+    }
+    const char *str = _PyUnicode_AsUTF8NoNUL(obj);
+    if (str == NULL) {
+        return NULL;
+    }
+    return str;
+}
+
+static PyObject *
+unicode_vectorcall(PyObject *type, PyObject *const *args,
+                   size_t nargsf, PyObject *kwnames)
+{
+    assert(Py_Is(_PyType_CAST(type), &PyUnicode_Type));
+
+    Py_ssize_t nargs = PyVectorcall_NARGS(nargsf);
+    if (kwnames != NULL && PyTuple_GET_SIZE(kwnames) != 0) {
+        // Fallback to unicode_new()
+        PyObject *tuple = _PyTuple_FromArray(args, nargs);
+        if (tuple == NULL) {
+            return NULL;
+        }
+        PyObject *dict = _PyStack_AsDict(args + nargs, kwnames);
+        if (dict == NULL) {
+            Py_DECREF(tuple);
+            return NULL;
+        }
+        PyObject *ret = unicode_new(_PyType_CAST(type), tuple, dict);
+        Py_DECREF(tuple);
+        Py_DECREF(dict);
+        return ret;
+    }
+    if (nargs > 3) {
```

Comment:
Can we use ths one?

```suggestion
    if (!_PyArg_CheckPositional("str", nargs, 0, 3)) {
        return NULL;
    }
```

Reply:
Well, it's more expensive; moreover there is precedent for this variant in the code base.

Reply:
I would not oppose it though.

Reply:
See https://github.com/python/cpython/pull/117746/files#r1560988495 for the actual suggestion.